net, http: allow setting socket connection timeout for http request

reneweb opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
http, net


##### Description of change

This allows passing the socket connection timeout to http#request
such that it will be set before the socket is connecting

Fixes #7580

cjihrig commented on lib/net.js:
```diff
@@ -67,7 +67,13 @@ exports.connect = exports.createConnection = function() {
+  if (args.socketTimeout) {
+    s.setTimeout(args.socketTimeout);
+    return Socket.prototype.connect.apply(s, args);
+  } else {
```
The `else` branch is unnecessary. You're returning the same thing in both branches.

reneweb replied:
oh yeah, will fix that

jasnell commented:
@nodejs/http

jasnell commented:
+1 to @cjihrig's comment. Once that is addressed this should be good.

mscdex commented on lib/_http_client.js:
```diff
@@ -133,10 +133,12 @@ function ClientRequest(options, cb) {
     self.shouldKeepAlive = false;
-    const optionsPath = {
-      path: self.socketPath
+    const createConnectionOptions = {
```
Why change the name?

reneweb replied:
It used to be `optionsPath` since the only option was the path. With this change we introduce non-path related options in there, which is why I felt renaming would make sense and I've chosen `createConnectionOptions` since we pass them to the `createConnection` function.

mscdex replied:
To me `optionsPath` is more general, like 'options for (unix) socket path' as opposed to 'options for TCP socket'. With that in mind, I'd prefer to see the name unchanged. If other @nodejs/collaborators disagree, at the very least I'd rather see a more succinct name like `connectOptions` or `createOptions` or similar.

mcollina replied:
I think a _shorter_ name is better.

reneweb replied:
I see what you mean @mscdex . I think it's ok to go back to `optionsPath` then.

mscdex commented on lib/_http_client.js:
```diff
@@ -133,10 +133,12 @@ function ClientRequest(options, cb) {
-      path: self.socketPath
+    const createConnectionOptions = {
+      path: self.socketPath,
+      socketTimeout: options.socketTimeout
```
Perhaps we can just use the name `timeout` instead of `socketTimeout`?

reneweb replied:
I think renaming to `timeout` would work. The main reason why I thought about naming it `socketTimeout` (or even `connectionTimeout`) is that we are also able to set the timeout via the `setTimeout` function after the `socket` \ `request` is created and as a user it wouldn't be immediately obvious why there are those two ways to provide a timeout. Hence, I tried to make it a little more clear with naming it `socketTimeout` instead of just `timeout`. Tbh, though, I am not sure if it really makes much of a difference.

mscdex replied:
The reason I bring it up is that passing an option named `socketTimeout` to the default `net.Socket` constructor seems redundant. Even though there are other 'socket' references when it comes to http.Agent, the fewer the better IMHO since third party Agent instances could be used which may not use sockets at all.

mcollina replied:
I agree with @mscdex, `timeout` sounds better.

reneweb replied:
OK, makes sense, will change to `timeout`

mcollina commented:
Some nits, but LGTM for me.

Trott commented:
Labeling `in progress` because it needs a test.

mscdex commented:
It also needs a documentation update

reneweb commented:
I've changed everything as discussed. I also noticed that we wouldn't emit a timeout event. To make it consistent with `setTimeout` - which does emit a timeout event - and to make it easier to test, I added that functionality. I've also added a test and updated the docs.

mscdex commented on lib/_http_client.js:
```diff
@@ -134,7 +135,8 @@ function ClientRequest(options, cb) {
     const optionsPath = {
-      path: self.socketPath
+      path: self.socketPath,
+      timeout: options.timeout
```
Minor nit, but to be consistent we could just reference `self.timeout` here instead.

reneweb replied:
Fixed it

jasnell commented on lib/_http_client.js:
```diff
@@ -559,6 +561,13 @@ function tickOnSocket(req, socket) {
+  };
+  if (req.timeout) {
+    socket.once('timeout', emitTimeout);
+  }
```
Why not simply:

```js
if (req.timeout)
  socket.once('timeout', () => req.emit('timeout'));
```

jasnell commented on test/parallel/test-http-client-timeout-option.js:
```diff
@@ -0,0 +1,41 @@
+'use strict';
+var common = require('../common');
+var assert = require('assert');
+var http = require('http');
```
Can you please make these `const`

jasnell commented on test/parallel/test-http-client-timeout-option.js:
```diff
@@ -0,0 +1,41 @@
+    // this space intentionally left blank
+  });
+  req.on('error', function() {
+    // this space is intentionally left blank
```
the error should not happen, correct? If so, I'd make this:
```js
req.on('error', common.fail);
```

reneweb replied:
It does throw a `socket hang up` error. I've checked other timeout related tests and they do that as well, so I guess it's ok.

jasnell commented on test/parallel/test-http-client-timeout-option.js:
```diff
@@ -0,0 +1,41 @@
+  });
+  req.on('close', function() {
+    server.close();
+  });
```
Please wrap this in a `common.mustCall()` ... like so:

```js
req.on('close', common.mustCall(() => server.close());
```

jasnell commented on test/parallel/test-http-client-timeout-option.js:
```diff
@@ -0,0 +1,41 @@
+  });
+
+  var timeout_events = 0;
+  req.on('timeout', function() {
```
`common.mustCall()` here as well

jasnell commented on test/parallel/test-http-client-timeout-option.js:
```diff
@@ -0,0 +1,41 @@
+  });
+  setTimeout(function() {
+    req.destroy();
+    assert.equal(timeout_events, 1);
```
Please use `assert.strictEqual()` here

jasnell commented:
Few additional nits. Almost there!

reneweb commented:
I've fixed all the mentioned issues, except for the req error handling one, which I commentent on. I've also applied the fixes to `test-http-client-timeout-event`, which I based my test off of, so it is consistent.

jasnell commented:
LGTM if CI is green.

santigimeno commented:
Shouldn't the `timeout` option be documented in the `net` module too?

reneweb commented:
True, I've added it now to the `net` documentation

imyller commented:
Looks like many of the requested changes have been made.

/ping @reneweb do you feel this is ready for landing?

imyller commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/4259/

imyller commented:
Removing `in-progress` label, because test has been implemented.

reneweb commented:
From the implementation point of view it should be fine. The only thing is that I got no feedback on, yet, are the doc changes, specifically for the `net` docs. Unless there are any objection there, it should be good to go.

imyller commented:
I'm counting three LGTMs and no clear objections for landing this. CI tests have passed.

I've assigned this to myself for landing prep, but I'd like to hold until monday to give everyone time to comment.

imyller commented:
I'll start landing this:

* Three LGTMs
* No objections
* All requested changes have been made
* CI tests passed (only known CI failures)

imyller commented:
landed in c9b59e8387b08dab20936f32d2b290804718f688

Thank you for your effort and contribution, @reneweb

bnoordhuis commented:
Can the people who authored/reviewed this comment on #12005?  It's unclear if the code actually does what the documentation suggests it does.

jasnell commented:
Ping @indutny @mcollina

mcollina commented:
@bnoordhuis to the best of my knowledge, it does what is supposed to be doing. The doc is definitely not clear in what is happening.
